### PR TITLE
feat #53 - Adding Client Flush API

### LINKED ISSUE #53: Add Flush API to Frisbee Client
The Frisbee client should be able to flush any queued messages whenever it wants to

### LINKED ISSUE #55: Close Error Channel
The error channel for async.errorCh should get closed once the goroutines for the async connection are closed

### LINKED ISSUE #56: Parallize Frisbee Test Cases
Frisbee test cases should all run in parallel

### PR DESCRIPTION
**Pull request by ShivanshVij**
## Description

Adding a `client.Flush` function to flush the underlying frisbee connection

Fixes #53 
Fixes #55 
Fixes #56 

## Type of change

**Please update the title of this PR to reflect the type of change.** (Delete the ones that aren't required).

- [ ] Bug fix (non-breaking change which fixes an issue) [title: 'fix:']
- [x] New feature (non-breaking change which adds functionality) [title: 'feat:']
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

N/A

## Benchmarking

Frisbee tries to adhere to strict performance requirements, so please make sure to run `go test -bench=. ./...` and paste the results below:

### Benchmarking Results:

```shell
goos: darwin
goarch: arm64
pkg: github.com/loopholelabs/frisbee
BenchmarkAsyncThroughputPipe32-10                     75          15962714 ns/op
BenchmarkAsyncThroughputPipe512-10                    50          23134186 ns/op
BenchmarkAsyncThroughputNetwork32-10                  96          13521316 ns/op
BenchmarkAsyncThroughputNetwork512-10                 55          24104079 ns/op
BenchmarkAsyncThroughputNetwork1024-10                30          38178347 ns/op
BenchmarkAsyncThroughputNetwork2048-10                20          56473256 ns/op
BenchmarkAsyncThroughputNetwork4096-10                12          95168750 ns/op
BenchmarkAsyncThroughputNetwork1mb-10                908           1210713 ns/op
BenchmarkClientThroughput/test-10                     51          26162674 ns/op
BenchmarkClientThroughputResponse/test-10             50          22758870 ns/op
BenchmarkThroughput/test-10                           40          28690877 ns/op
BenchmarkThroughputWithResponse/test-10               49          27220836 ns/op
BenchmarkSyncThroughputPipe32-10                       6         167798930 ns/op
BenchmarkSyncThroughputPipe512-10                      6         172312417 ns/op
BenchmarkSyncThroughputNetwork32-10                    2         511359500 ns/op
BenchmarkSyncThroughputNetwork512-10                   3         501820472 ns/op
BenchmarkSyncThroughputNetwork1024-10                  3         439962042 ns/op
BenchmarkSyncThroughputNetwork2048-10                  3         433430528 ns/op
BenchmarkSyncThroughputNetwork4096-10                  3         416769028 ns/op
BenchmarkSyncThroughputNetwork1mb-10                1298            934396 ns/op
PASS
ok      github.com/loopholelabs/frisbee 52.968s
?       github.com/loopholelabs/frisbee/internal/errors [no test files]
goos: darwin
goarch: arm64
pkg: github.com/loopholelabs/frisbee/internal/protocol
BenchmarkEncodeHandler-10               75834147                15.30 ns/op
BenchmarkDecodeHandler-10               98391907                12.22 ns/op
BenchmarkEncodeDecodeHandler-10         41406616                28.86 ns/op
BenchmarkEncode-10                      96218098                12.38 ns/op
BenchmarkDecode-10                      135815473                8.715 ns/op
BenchmarkEncodeDecode-10                50067799                23.66 ns/op
PASS
ok      github.com/loopholelabs/frisbee/internal/protocol       8.936s
PASS
ok      github.com/loopholelabs/frisbee/internal/ringbuffer     0.091s
```

## Linting

Please make sure you've run the following and fixed any issues that arise:

- [x] `go fmt ./...` has been run
- [x] `golangci-lint run --skip-dirs-use-default` has been run

## Final Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: async_test.go [hunk 1]
```diff
@@ -315,7 +330,7 @@ func TestAsyncTimeout(t *testing.T) {
 	assert.Equal(t, *message, *readMessage)
 	assert.Nil(t, content)
 
-	time.Sleep(defaultDeadline * 10)
+	time.Sleep(defaultDeadline * 5)
```
Comment (Jspsun) [inline async_test.go:333]: is there a reason why we changed this? Just improve test times?

File: async_test.go [hunk 2]
```diff
@@ -547,14 +562,14 @@ func BenchmarkAsyncThroughputNetwork1024(b *testing.B) {
 	var reader, writer net.Conn
 	start := make(chan struct{}, 1)
 
-	l, _ := net.Listen("tcp", ":3000")
+	l, _ := net.Listen("tcp", ":")
```
Comment (Jspsun) [inline async_test.go:565]: I think it wouldnt make a difference but we shoudl explicitly have this as ":0"

File: client.go [hunk 3]
```diff
@@ -133,9 +133,14 @@ func (c *Client) WriteMessage(message *Message, content *[]byte) error {
...
 
-// ErrorChannel returns an error channel that can be listened to see if this client has an error
-func (c *Client) ErrorChannel() <-chan error {
-	return c.conn.ErrorChannel()
+// Flush flushes any queued frisbee Messages from the client to the server
+func (c *Client) Flush() error {
+	return c.conn.Flush()
+}
+
+// CloseChannel returns an channel that can be listened to see if this client has been closed
```
Comment (Jspsun) [inline client.go:141]: nit: "an" channel

File: conn.go [hunk 4]
```diff
@@ -30,19 +30,11 @@ import (
...
-
-	// CLOSED is used to specify that the connection has been closed (possibly due to an error)
-	CLOSED
-)
-
 var (
 	defaultLogger   = zerolog.New(os.Stdout)
 	defaultDeadline = time.Second
 	emptyTime       = time.Time{}
+	pastTime        = time.Unix(1, 0)
```
Comment (Jspsun) [inline conn.go:37]: Why do we have this? I dont see anwhere that we are using this

**Review by Jspsun — DISMISSED** [review]

File: async_test.go (hunk 1 above)
Comment (ShivanshVij) [reply async_test.go:333]: Yep

File: conn.go (hunk 4 above)
Comment (ShivanshVij) [reply conn.go:37]: We're using it in the `killGoroutines` function in `async.go`

File: async_test.go (hunk 2 above)
Comment (ShivanshVij) [reply async_test.go]: Fixed

File: client.go (hunk 3 above)
Comment (ShivanshVij) [reply client.go]: Fixed

**Pull request merged**